src/ast.rs

```diff
@@ -89,11 +99,69 @@ impl UnionElement {
     pub fn get<'a>(&self, v: &'a Value) -> Iter<'a> {
         match self {
             UnionElement::Name(name) => Box::new(v.get(name).into_iter()),
+            UnionElement::Slice(slice) => {
+                if let Value::Array(arr) = v {
+                    let step = slice.step.unwrap_or(1);
+
+                    let len = arr.len();
+
+                    let start = slice
+                        .start
+                        .map(|s| if s < 0 { s + (len as isize) } else { s })
+                        .unwrap_or(if step > 0 { 0 } else { (len as isize) - 1 });
+
+                    let end = slice
+                        .end
+                        .map(|e| if e < 0 { e + (len as isize) } else { e })
+                        .unwrap_or(if step > 0 { len as isize } else { -1 });
+
+                    Box::new(array_slice(arr, start, end, step))
+                } else {
+                    Box::new(iter::empty())
+                }
+            }
             UnionElement::Index(num) => Box::new(v.get(abs_index(*num, v)).into_iter()),
         }
     }
 }
 
+fn array_slice(arr: &[Value], start: isize, end: isize, step: isize) -> Iter<'_> {
+    let len = arr.len();
+    let mut sl = vec![];
+    match step.cmp(&0) {
+        Ordering::Greater => {
+            let strt = if start < 0 { 0 } else { start as usize }; // avoid CPU attack
+            let e = if end > (len as isize) {
+                len
+            } else {
+                end as usize
+            }; // avoid CPU attack
+            for i in (strt..e).step_by(step as usize) {
+                if i < len {
+                    sl.push(&arr[i]);
```

Comment:
The spec says:

```
A negative index j selects an element of an array of length len if and only if 0 <= j + len < len, in which case it selects the same element as the non-negative index j + len.
```

then it describes the iterations as:

```
for (i = start; i < end; i = i + step) {
  ...
}
```

what should happen if you call this on an input slice of length 5?:
```rust
    let s = array_slice(j.as_array().unwrap(), -5, 0, 1);
```

My interpretation is that it would yield

```
i = -5;
i < 0 //  true
push(arr[i + len]) // arr[-5+5] === arr[0]
i = i+1 //  -4
i < 0 //  true
push(arr[i + len]) // arr[-4+5] === arr[1]
...
```

Reply:
This is a good example of a reference implementation and compliance test suite turning up a hole in the spec.

I think the spec _should_ say (only more formally/precisely) that negative `start` is syntactic sugar for `start + len` and similarly for negative `end`. But then `start` and `end` need to be desugared _before_ plugging them into the relevant `for` loop.

So, for an array of length 5, `[-5, 0, 1]` corresponds to the elements indexed by the values of `i` in the `for` loop:
```
for (i = 0; i < 0; i = i + 1) {
  ...
}
```
of which there are none, so the result is empty.

Reply:
Desugaring before iterating makes most sense to me as well.

A way tinkering with the consequences if doing it the other way around and I don't think it would provide any advantage. An interesting consequence is that it allows the resulting slice to be longer than the input array (and contain repeated elements) which I think would be quite confusing.

Reply:
I think the implementation is correct and we just need to fix the spec. ;-)

Reply:
the privileges of a spec writer! :-)

Reply:
I wonder if something like this would have less special casing: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dd7180d6586e5a511b4e56b0058e5562